File: 107-binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void reverse(vector<vector<int>> &t){
        int i=0, j=t.size()-1;
        while(i < j){
            swap(t[i], t[j]);
            i++; j--;
        }
    }
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> traversal;
        if (root == nullptr) return traversal;

        queue<TreeNode*> q;
        q.push(root);

        while(!q.empty()){
            int n = q.size();
            vector<int> level(n);
            for(int i=0; i<n; i++){
                TreeNode *node = q.front();
                q.pop();

                level[i] = node->val;

                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            traversal.push_back(level);
        }

        reverse(traversal);

        return traversal;
    }
};

File: 107-binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrderBottom(root *TreeNode) [][]int {
    var traversal [][]int
    if root == nil {
        return traversal
    }

    var q []*TreeNode
    q = append(q, root)

    for len(q) != 0 {
        n := len(q)
        var level []int
        for i := 0; i < n; i++ {
            node := q[0]
            q = q[1:]

            level = append(level, node.Val)

            if node.Left != nil { q = append(q, node.Left) }
            if node.Right != nil { q = append(q, node.Right) }
        }

        traversal = append(traversal, level)
    }

    reverse(traversal)

    return traversal
}
func reverse(nums [][]int) {
    i, j := 0, len(nums)-1

    for i < j {
        nums[i], nums[j] = nums[j], nums[i]
        i++
        j--
    }
}